### PR TITLE
Add OpmCuILU0 and clean up cuistl

### PR DESCRIPTION
**Pull request by multitalentloes**
This PR adds our own implementation of ILU0 in cuda using an implementation similar to that of CuDILU.

SPE11C speedup on GPU vs 16 processes on AMD Ryzen 5950x, both running ILU0 preconditioner:
The comparison is the time spent in the linear solver normalized with the respect to the number of linear iterations, meaning time spent moving data up and down from the GPU is included included.
| num of active cells | AMD Radeon RX 7900XTX Speedup | Nvidia RTX 4070Ti Speedup |
|--------------|------------------------------|---------------------------|
| 35k        | 0.21 | 0.34 |
| 115k       | 0.91 | 1.20 |
| 270k       | 1.61 | 1.99 |
| 903k       | 3.31 | 3.89 |
| 2.13m      | 4.20 | 4.71 |
| 3.09m      | 4.76 | 5.59 |

The PR also cleanes up some files like cusparsematrixoperations that were growing too large. A new folder with separate files for the kernels of the different preconditioners is created.

A following PR (#5451) will avoid the introduced code duplication of the thread block tuning step that for now is present in both the CuDILU and OpmCuILU0 cpp files and improve it with cuda events.

### CODE REVIEW

**Comment by multitalentloes** [comment]
Jenkins build this please

**Comment by kjetilly** [comment]
Jenkins build this please

**Review by kjetilly — CHANGES_REQUESTED** [review]
In general no large comment, but minor fixes here and there, most notably unused includes. Probably a good idea to run clang_format on all files.

File: opm/simulators/linalg/cuistl/CuDILU.cpp [hunk 1]
```diff
@@ -22,113 +22,29 @@
...
 #include <fmt/core.h>
 #include <opm/common/ErrorMacros.hpp>
+#include <opm/simulators/linalg/cuistl/detail/coloringAndReorderingUtils.hpp>
 #include <opm/simulators/linalg/cuistl/CuDILU.hpp>
 #include <opm/simulators/linalg/cuistl/CuSparseMatrix.hpp>
 #include <opm/simulators/linalg/cuistl/CuVector.hpp>
 #include <opm/simulators/linalg/cuistl/detail/cuda_safe_call.hpp>
 #include <opm/simulators/linalg/cuistl/detail/cusparse_matrix_operations.hpp>
+#include <opm/simulators/linalg/cuistl/detail/preconditionerKernels/DILUKernels.hpp>
 #include <opm/simulators/linalg/cuistl/detail/safe_conversion.hpp>
```
[the remark this answers is not in the document]
Comment (multitalentloes) [reply opm/simulators/linalg/cuistl/CuDILU.cpp:32]: The two includes pointed out in this file are in use, but clang-tidy have pointed out some others that I will remove

File: opm/simulators/linalg/cuistl/CuDILU.cpp [hunk 2]
```diff
@@ -155,10 +71,10 @@ CuDILU<M, X, Y, l>::CuDILU(const M& A, bool splitMatrix, bool tuneKernels)
                              A.nonzeroes()));
     if (m_splitMatrix) {
         m_gpuMatrixReorderedDiag.reset(new auto(CuVector<field_type>(blocksize_ * blocksize_ * m_cpuMatrix.N())));
```
[the remark this answers is not in the document]
Comment (multitalentloes) [reply opm/simulators/linalg/cuistl/CuDILU.cpp:73]: rewritten to just be a make_unique now

**Comment by multitalentloes** [comment]
Comments have now been addressed and commits are squashed

**Comment by multitalentloes** [comment]
Jenkins build this please

**Review by kjetilly — APPROVED** [review]

**Pull request merged**